Feature metered connection capability

nucleogenesis opened the pull request:
<!--
 1. Following guidance below, replace …'s with your own words
 2. After saving the PR, tick of completed checklist items
 3. Skip checklist items that are not applicable or not necessary
 4. Delete instruction/comment blocks
-->

## Summary
<!--
 * description of the change
 * manual verification steps performed
 * screenshots if the PR affects the UI
-->

**Completes the implementation of the `MeteredConnectionModal` component in #9779** 

- Wires up new capability `CHECK_IS_METERED` to indicate whether the app  is able to get the value from the device
- Updates the `integration_testing/scripts/run_kolibri_app_mode.py` script to register the new capability and return `True` by default (devs can just change it to whatever they want when they want during testing locally)
- Includes the `MeteredConnectionModal` in the `CorePage/AppBarPage` component
- Updates the `MeteredConnectionModal` to
    - Fetch the user's device setting
    - Decide whether to show itself if:
        - The user is a superuser
        - The user's setting is to _not allow_ a metered connection
        - The user is currently trying to use a metered connection
        - The user has not previously answered this modal
    - Update the `DeviceSettings.extra_settings.allow_download_on_metered_connection` based on the user's choice
    - Maintain a `sessionStorage` value indicating that the modal has been dismissed before

**Some remaining thoughts / questions**

- Does this also need to be hooked up to the `DeviceSettings` page - I've not done that here
- :question:  - Should this instead be something we trigger - ie, whenever the user goes to make a download or should it just be shown no matter what when there is a metered connection detected for the first time?
- We ought to show a snackbar message when the user's settings are updated so I have added the code for it, but I've commented it out and will make a follow-up issue for it for after the release once this merges
 

## References
<!--
 * references to related issues and PRs
 * links to mockups or specs for new features
 * links to the diffs for any dependency updates, e.g. in iceqube or the perseus plugin
-->

#9779 

## Reviewer guidance
<!--
 * how can a reviewer test these changes?
 * are there any risky areas that deserve extra testing
-->

### Testing the APK (@pcenov)

[<<  LINK TO APK >>
](https://transfer.sh/rE5aVU/kolibri-debug-0.16.0a13.dev0+git.23.g3f716578-4235453-unknown.apk)
You should be able to install this on your device. On mine I am able to go into the network settings and override my connection to force it to be treated as "metered". 

Once you install the APK and go through the setup wizard you should see the pop-up if you're on a metered connection. Navigate around a bit and you should basically never see it again.

To reset the state and test again I think you'll have to reinstall the APK or manually clear all data for the app in Android settings for the app.

### Testing as a dev

Run `yarn app-devserver` to run Kolibri in the app mode. Be sure to click the `initialize/` url that will come up in your console. If you want to test with the connection not being metered, update the `run_kolibri_app_mode.py` file's `check_is_metered` function to return `False` and restart the `yarn app-devserver`.

Log in as a superuser and you should see the modal the first time and never again after. The key is stored in session storage, clear it and you'll see it again if you selected to not allow connections. As is, if you click "Allow..." then you'll never see the modal again no matter what per the expected logic of the modal not showing whenever they've already said it was allowed. Until the Device Settings page is updated you won't be able to change this directly without going into the Kolibri shell and running:

```python
from kolibri.core.device.models import DeviceSettings
settings = DeviceSettings.objects.first()
ex = settings.extra_settings
ex["allow_download_on_metered_connection"] = False
settings.extra_settings = ex
settings.save()
```

### Code Review Thoughts / Questions

- Greatly appreciate any thoughts / suggestions!
- I added `metered` as a value in the Vuex `core/connection` namespace. It felt like it was a fine place to put it, but perhaps we can start migrating some of our Vuex modules like these to a `useNetwork` module. 
- If this needs to be triggered during a particular workflow, instead of right when the user signs up, then I can work out a way to do that - maybe should just move it out of the `<PageRoot>` altogether and leave it to whatever component that needs it to include it? Just not sure where that should be.

----

## Testing checklist

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests


## PR process

- [ ] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

nucleogenesis committed:
add CHECK_IS_METERED capability

nucleogenesis committed:
wire things up

nucleogenesis committed:
okay now we see if we are metered

nucleogenesis committed:
debug log better

nucleogenesis committed:
fixerupper, simpler appCapability fn

nucleogenesis committed:
metered status setup in core/connection; core-app/index;

nucleogenesis committed:
metered connection polling all in Metered...vue file now, logic for display and dismissal cookie

nucleogenesis committed:
improve comment

nucleogenesis committed:
move updateMeteredStatus to core.actions namespace

rtibbles commented on kolibri/core/assets/src/state/modules/connection.js:
```diff
@@ -2,8 +2,12 @@ export default {
   state: {
```
I'd rather not have this live in here, mostly because the `connection` this vuex module refers to is the one between the browser and the backend, which is distinctly different from the one we are concerned with whether it is metered.

Maybe let's just move this out into a nice little composable all of its own?

rtibbles commented on kolibri/core/assets/src/utils/appCapabilities.js:
```diff
@@ -18,6 +18,23 @@ export const checkCapability = key => store.getters.isAppContext && appCapabilit
+      logging.warn('Checking if the device is metered is not supported on this platform');
+      return Promise.reject();
+    }
+    return client({ url: urlFunction(), method: 'GET' });
```
Could we maybe do:
```suggestion
    return client({ url: urlFunction(), method: 'GET' }).then(response => response.data.value);
```

rtibbles commented on kolibri/core/assets/src/utils/appCapabilities.js:
```diff
@@ -18,6 +18,23 @@ export const checkCapability = key => store.getters.isAppContext && appCapabilit
+   */
+  checkIsMetered() {
+    if (!checkCapability('check_is_metered')) {
+      return Promise.resolve({ data: { value: null } });
```
With the change below this could just be:
```suggestion
      return Promise.resolve(null);
```

rtibbles commented on kolibri/core/assets/src/utils/appCapabilities.js:
```diff
@@ -18,6 +18,23 @@ export const checkCapability = key => store.getters.isAppContext && appCapabilit
+    const urlFunction = urls['kolibri:kolibri.plugins.app:appcommands-check-is-metered'];
+    if (!urlFunction) {
+      logging.warn('Checking if the device is metered is not supported on this platform');
+      return Promise.reject();
```
This would happen if the capability was registered, but the app plugin was not enabled. Either way, I feel like it would be better to just resolve to `null` again like above?

So could maybe look for the URL function first, then just check:
```
if (!urlFunction || !checkCapability('check_is_metered')) {
```

rtibbles commented on packages/kolibri-common/components/MeteredConnectionNotificationModal.vue:
```diff
@@ -28,26 +32,106 @@
+      // otherwise this component will never be rendered anyway so no need to make the calls
+      if (checkCapability('check_is_metered')) {
+        // Get status initially and poll every 5 seconds
+        this.$store.dispatch('updateMeteredStatus').then(() => {
```
As you suggest in the question in your PR description, I don't think we should be polling this - only triggering the modal when the user initiates an action that might lead to a download.

marcellamaki replied:
yes - I agree with this. I don't think we explicitly wrote this anywhere, but it seems like the better approach

rtibbles commented on kolibri/core/assets/src/views/CorePage/AppBarPage.vue:
```diff
@@ -44,6 +44,8 @@
       @cancel="languageModalShown = false"
     />
 
+    <MeteredConnectionNotificationModal />
```
With the explicit trigger in mind, we also don't need to add this here. Just in places where a superuser initiates a download.

marcellamaki replied:
@nucleogenesis let me know if it would be helpful for me to audit all of these locations

rtibbles reviewed (dismissed):
Needs some tweaks, but the main core of the backend and frontend data interaction is here. @marcellamaki do we have an issue covering how this is connected to user interactions based on the figma spec? I feel like that would help guide @nucleogenesis's follow up here.

pcenov commented:
@nucleogenesis the link to the .apk file is broken - I guess you can send it to me over Slack?

pcenov commented:
Hi @nucleogenesis, I confirm that I am able to see the modal if I've set the network to 'metered'.
At the same time I am observing the following issues:
1. If I select the option 'Do not allow Kolibri to use mobile data' then the modal keeps reappearing each time I relaunch the app.
2. If I select the 'Allow Kolibri to use the mobile data' option then the modal doesn't reappear when I relaunch the app but it does appear briefly if I navigate to 'Device settings' with both options disabled and curiously enough 'Do not allow Kolibri to use mobile data' is the selected one.

Here's a video for that:

https://user-images.githubusercontent.com/79847249/236198455-a455897f-f404-42c7-a585-05238cfcd7e3.mp4

I am assuming that the scope of the testing for this PR should be limited to your specific instructions about when the modal should appear and when it should not. Let me know if you need anything else tested here.

marcellamaki commented:
@rtibbles - we don't, just we had some conversations about having it at the "point of download" or another checkpoint. I can write up something for this, if that would be helpful? @nucleogenesis  -- either for a follow up issue or here

nucleogenesis committed:
strip out unnecessary complexity, prepare for triggerable

nucleogenesis commented:
@rtibbles @marcellamaki

I've updated the code and removed the Vuex business. When I removed it I realized that it was only really useful when we're polling for the metered connection status.

Since we're going to trigger it -- we can just "trigger" the `<MeteredConnectionModal />` with a `v-if` and then the modal, when it mounts, will check if it should actually be shown (ie, isSuperuser, device setting is to disallow, etc).

---

@pcenov 

re: The modal showing after dismissal saying "do not allow":
It seems like we cannot rely on the `sessionStorage` as I'd hoped. I'll make a follow-up issue to this one to figure out if maybe we'll need to set the value on the user's session in Django (cc @rtibbles ?)

re: The brief flash
This should be resolved naturally once we start only showing it on a particular trigger. I think that the issue was related to the fact that the component itself was remaining mounted, albeit hidden, so its state wasn't being reset. When the component is setup to only show in particular user flows, then I think this will no longer be an issue.

marcellamaki commented:
@nucleogenesis - I've added [a follow up issue ](https://github.com/learningequality/kolibri/issues/10662) for the places the metered connection modal should appear

nucleogenesis commented:
Just a note that I think that this should be good on Code Review - I'll add a note on #10662 to also remove the modal from AppBarPage when that work is done as well.

@pcenov I'll ping you for proper UX review on the follow-up PR as that will be the proper implementation of the modal while this was just added as a way to test the underlying architecture.

akolson reviewed (commented):
Thanks @nucleogenesis. Code makes sense to me. ~~Looks like Richard left a comment about notification triggering that might need to be implemented as well~~ after reading comment above.

nucleogenesis commented:
I've marked this as ready for review - #10662 has all necessary follow-up tasks noted there to complete this feature so this PR ought to be ready to merge.

marcellamaki reviewed (approved):
It looks like all of Richard's feedback has been addressed here, and that the to-dos in the follow up issue cover what remains. I added one more item to clean up some of the console logs, as I think leaving them for now is probably useful for finalizing, but since they will already be merged, they probably won't be caught in code review the next time.
